src/roofs.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+#ifndef ROOFS_H
+#define ROOFS_H
+
+#include <map>
+#include <memory>
+#include "string_id.h"
+
+class map;
+class game;
+struct tripoint;
+
+struct support_cache;
+
+struct ter_t;
+using ter_id = int_id<ter_t>;
+
+enum direction : unsigned;
+
+enum class roof_dir : size_t {
+    west = 0,
+    north,
+    east,
+    south,
+    num_dirs
+};
+
+class roofs
+{
+    private:
+        game &g;
+        // @todo Roofs should collapse even outside bubble, so map dependency is wrong
+        map &m;
+
+        std::unique_ptr<support_cache> cache;
+    public:
+        roofs( game &gm, map &mp );
```

Comment:
I would like for this guy to not have a dependency on game or map, instead pass in a std::unordered_map<tripoint, support_struct> or similar to do your calculations on, and then ask this guy things about it.  Then we could write some unit tests with synthetic data to make assertions about some common scenarios.